typescript/agentkit/src/wallet-providers/smartWalletProvider.ts

```diff
@@ -0,0 +1,366 @@
+import {
+  CHAIN_ID_TO_NETWORK_ID,
+  Coinbase,
+  createSmartWallet,
+  NetworkScopedSmartWallet,
+  SendUserOperationOptions,
+  Signer,
+  SupportedChainId,
+  toSmartWallet,
+  waitForUserOperation,
+} from "@coinbase/coinbase-sdk";
+import {
+  Abi,
+  Address,
+  ContractFunctionArgs,
+  ContractFunctionName,
+  createPublicClient,
+  Hex,
+  http,
+  Prettify,
+  ReadContractParameters,
+  ReadContractReturnType,
+  TransactionRequest,
+  PublicClient as ViemPublicClient,
+} from "viem";
+import { Network, NETWORK_ID_TO_CHAIN_ID, NETWORK_ID_TO_VIEM_CHAIN } from "../network";
+import { EvmWalletProvider } from "./evmWalletProvider";
+import { version } from "../../package.json";
+
+export interface ConfigureSmartWalletOptions {
+  cdpApiKeyName?: string;
+  cdpApiKeyPrivateKey?: string;
+  networkId?: string;
+  smartWalletAddress?: Hex;
+  paymasterUrl?: string;
+  signer: Signer;
+}
+
+interface SmartWalletProviderConfig {
+  smartWallet: NetworkScopedSmartWallet;
+  network: Required<Network>;
+  chainId: string;
+}
+
+/**
+ * A wallet provider that uses Smart Wallets from the Coinbase SDK.
+ */
+export class SmartWalletProvider extends EvmWalletProvider {
+  #smartWallet: NetworkScopedSmartWallet;
+  #network: Required<Network>;
+  #publicClient: ViemPublicClient;
+
+  /**
+   * Constructs a new CdpWalletProvider.
+   *
+   * @param config - The configuration options for the CdpWalletProvider.
+   */
+  private constructor(config: SmartWalletProviderConfig) {
+    super();
+
+    this.#network = config.network;
+    this.#smartWallet = config.smartWallet;
+    this.#publicClient = createPublicClient({
+      chain: NETWORK_ID_TO_VIEM_CHAIN[config.network.networkId],
+      transport: http(),
+    });
+  }
+
+  /**
+   * Configures and returns a `SmartWalletProvider` instance using the provided configuration options.
+   * This method initializes a smart wallet based on the given network and credentials.
+   *
+   * @param {ConfigureSmartWalletOptions} config
+   *   - Configuration parameters for setting up the smart wallet.
+   *
+   * @returns {Promise<SmartWalletProvider>}
+   *   - A promise that resolves to an instance of `SmartWalletProvider` configured with the provided settings.
+   *
+   * @throws {Error}
+   *   - If an invalid combination of `networkId` and `chainId` is provided.
```

Comment:
I think this one is also no longer accurate, or at least I am not seeing it

Reply:
Yeah the whole error is no longer accurate. After the refactor, the only real error here is passing in a unsupported network ID (whether because its not in AgentKit's `NETWORK_ID_TO_CHAIN_ID` or if its not in `@coinbase/coinbase-sdk`'s `CHAIN_ID_TO_NETWORK_ID`